README.md
# cointmonitoR

Consistent Monitoring of Stationarity and Cointegrating Relationships.

[![Travis-CI Build Status](https://travis-ci.org/aschersleben/cointmonitoR.svg?branch=master)](https://travis-ci.org/aschersleben/cointmonitoR)
[![CRAN Status Badge](http://www.r-pkg.org/badges/version/cointmonitoR)](https://cran.r-project.org/package=cointmonitoR) 

* Installation (including latest CRAN version of `cointReg`)
```r
install.packages("cointmonitoR")
library("cointmonitoR")
```

* Simple example (stationarity, structural break):
```r
set.seed(1909)
eps <- rnorm(200)
x <- c(eps[1:100], cumsum(eps[101:200]) / 2)
test <- monitorStationarity(x, m = 93)
print(test)
oldpar <- par(mfrow = c(2, 1))
plot(test, what = "both", legend = FALSE)
par(oldpar)
```

* Package vignette: Provides further examples and explanations.
```r
vignette("cointmonitoR")
```

* Package help page: Overview of all available functions:
```r
package?cointmonitoR
```

* [Package issues](https://github.com/aschersleben/cointmonitoR/issues)

* [Package NEWS](https://github.com/aschersleben/cointmonitoR/blob/master/inst/NEWS.md)

* Installation of the current development version from GitHub:
```r
devtools::install_github("aschersleben/cointmonitoR", build_vignettes = TRUE)
```


## Theoretical background

We propose a consistent monitoring procedure to detect a structural change from a cointegrating relationship to a spurious relationship. The procedure is based on residuals from modified least squares estimation, using either Fully Modified, Dynamic or Integrated Modified OLS. It is inspired by [Chu et al. (1996)](http://dx.doi.org/10.2307/2171955) in that it is based on parameter estimation on a pre-break "calibration" period only, rather than being based on sequential estimation over the full sample.

See the [discussion paper](http://dx.doi.org/10.2139/ssrn.2624657) for further information.

This package provides the monitoring procedures for both the cointegration and the stationarity case (the latter is just a special case of the former one) as well as printing and plotting methods for a clear presentation of the results.


## References

* Aschersleben, P. and M. Wagner (2016). cointReg: Parameter Estimation and Inference in a Cointegrating Regression. R package version 0.2.0. https://CRAN.R-project.org/package=cointReg

* Chu, C.J., M. Stinchcombe and H. White (1996): "Monitoring Structural Change", _Econometrica_, 64, 1045--1065, [DOI:10.2307/2297912](http://dx.doi.org/10.2307/2297912).

* Wagner, M. and D. Wied (2015): "Monitoring Stationarity and Cointegration," _Discussion Paper_, [DOI:10.2139/ssrn.2624657](http://dx.doi.org/10.2139/ssrn.2624657)


inst/NEWS.md
# cointmonitoR dev
 * Fix some formats and links.

# cointmonitoR 0.1.0 (2016-06-14)
 * First CRAN release version of the package.


R/plot.R
#' Plot Method for Monitoring Procedures.
#'
#' Plotting objects of class \code{"cointmonitoR"}.
#'
#' @param x [\code{cointmonitoR}]\cr
#'   Object of class \code{"cointmonitoR"}, i.e. the result of
#'   \code{\link{monitorStationarity}} or
#'   \code{\link{monitorCointegration}}.
#'
#' @param what [\code{character}]\cr
#'   Whether to plot test statistics (\code{"test"}) (default) or the
#'   values/residuals of the tested time series (\code{"values"} or
#'   \code{"residuals"}) or \code{"both"}. Works only,
#'   if \code{return.stats = TRUE} in the called function that to get \code{x}
#'   (default setting).
#'
#' @param type [\code{character}]\cr
#'   Plot type (from \code{\link{plot}}). Default is \code{"l"}.
#'
#' @param main,xlab,ylab [\code{character}]\cr
#'   Title and axis titles (from \code{\link{plot}}). Default values will be
#'   generated from the contents of \code{x}.
#'
#' @param axes,legend [\code{logical}]\cr
#'   Whether to add axes (from \code{\link{plot}}) and a legend to the plot.
#'
#' @param main.val,xlab.val,ylab.val [\code{character}]\cr
#'   Title and axis titles (from \code{\link{plot}}) for the second plot,
#'   if generating both plots in one step (see argument \code{what}).
#'   Default values will be generated from the contents of \code{x}.
#'
#' @param lines [\code{logical}]\cr
#'   Whether to add lines and annotations to the plot. Default is \code{TRUE}.
#'
#' @param ... [\code{any}]\cr
#'   Further arguments passed to \code{\link{plot}}.
#'
#' @family cointmonitoR
#'
#' @examples
#' ### Monitoring stationarity (no break):
#' set.seed(1909)
#' x = rnorm(200)
#' test = monitorStationarity(x, m = 0.5)
#' plot(test)
#'
#' oldpar = par(mfrow = c(2, 1), mar = c(4, 4, 1, 1))
#' plot(test, what = "both", legend = FALSE, main = "", main.val = "")
#' par(oldpar)
#'
#'
#' ### Monitoring stationarity (break):
#' x = c(x[1:100], cumsum(rnorm(100, sd = 0.5)) + x[101:200])
#' test2 = monitorStationarity(x, m = 0.5)
#' plot(test2)
#'
#' oldpar = par(mfrow = c(2, 1), mar = c(4, 4, 1, 1))
#' plot(test2, what = "both", legend = FALSE, main = "", main.val = "")
#' par(oldpar)
#'
#'
#' ### Monitoring cointegration (no break):
#' set.seed(42)
#' x = data.frame(x1 = cumsum(rnorm(200)), x2 = cumsum(rnorm(200)))
#' eps1 = rnorm(200, sd = 2)
#' y = x$x1 - x$x2 + 10 + eps1
#' test3 = monitorCointegration(x = x, y = y, m = 0.5, model = "FM")
#' plot(test3)
#'
#' oldpar = par(mfrow = c(2, 1), mar = c(4, 4, 1, 1))
#' plot(test3, what = "both", legend = FALSE, main = "", main.val = "")
#' par(oldpar)
#'
#'
#' ### Monitoring cointegration (break):
#' eps2 = c(eps1[1:100], cumsum(eps1[101:200]))
#' y = x$x1 - x$x2 + 10 + eps2
#' test4 = monitorCointegration(x = x, y = y, m = 0.5, model = "FM")
#' plot(test4)
#'
#' oldpar = par(mfrow = c(2, 1), mar = c(4, 4, 1, 1))
#' plot(test4, what = "both", legend = FALSE, main = "", main.val = "")
#' par(oldpar)
#'
#' @importFrom graphics plot.default abline text legend par
#'
#' @export


plot.cointmonitoR <- function(x, what = "test", type, main, xlab, ylab,
                              axes = TRUE, legend = TRUE,
                              main.val, xlab.val, ylab.val, lines = TRUE, ...) {

  if (!(class(x) == "cointmonitoR"))
    stop("Argument x must be of type \"cointmonitoR\".")

  firstToCap <- function(x) paste0(toupper(substr(x, 1, 1)), substring(x, 2))

  what <- match.arg(what, c("test", "values", "residuals", "both"),
                    several.ok = TRUE)
  if ("both" %in% what) {
    what <- c("test", "values")
  }

  calib.col <- 4
  calib.lty <- 2

  crit.col <- "grey"
  crit.lty <- 2

  break.col <- 2
  break.lty <- 1

  if (missing(type))
    type <- "l"

  mon.trend <- firstToCap(x$trend)
  mon.type <- firstToCap(attr(x, "type"))

  if (missing(main))
    main <- paste("Monitoring Procedure for", mon.trend, mon.type)

  if (missing(xlab))
    xlab <- "Observation Number"

  if ("test" %in% what) {
    if (missing(ylab))
      ylab <- paste("Test Statistics (Hsm)")

    plot.default(x$statistics, type = type, main = main, xlab = xlab,
                 ylab = ylab, axes = axes, ...)
    # points(c(0, x$m$m.index), c(0, 0), type = "b", col = 2, pch = 19)
    if (lines) abline(v = c(0, x$m$m.index), col = calib.col, lty = calib.lty)
    # text(x$m$m.index / 2, 0, "Calibration", pos = 3, col = 2)

    if (x$p.value <= x$sig) {
      if (lines) {
        abline(h = x$cv, col = crit.col, lty = crit.lty)
        abline(v = x$time, col = break.col, lty = break.lty)
      }
      if (legend) {
        legend("topleft", col = c(calib.col, crit.col, break.col),
               lty = c(calib.lty, crit.lty, break.lty),
               legend = c("Calibration Period", "Critical Value",
                          paste0("Detected Break Point (", x$time, ")")))
      }
    } else {
      if (lines) {
        text(x$m$m.index / 2, par("usr")[3], "Calibration", pos = 3,
             col = calib.col, lty = calib.lty)
      }
    }
  }

  if ("values" %in% what | "residuals" %in% what) {
    if (missing(ylab.val)) {
      if ("test" %in% what || missing(ylab))
        ylab.val <- x$name
      else ylab.val <- ylab
    }

    if (mon.type == "Stationarity") {
      plot.default(as.numeric(x$input), type = type, main = main, xlab = xlab,
                   ylab = ylab.val, ...)
    } else {
      plot.default(x$residuals, type = type, main = main, xlab = xlab,
                   ylab = ylab.val, ...)
    }

    if (lines) abline(v = c(0, x$m$m.index), col = calib.col, lty = calib.lty)

    if (x$p.value <= x$sig) {
      if (lines) abline(v = x$time, col = break.col, lty = break.lty)
      if (legend) {
        legend("topleft", col = c(calib.col, break.col),
               lty = c(calib.lty, break.lty),
               legend = c("Calibration Period",
                          paste0("Detected Break Point (", x$time, ")")))
      }
    } else {
      if (lines) {
        text(x$m$m.index / 2, par("usr")[3], "Calibration", pos = 3,
             col = calib.col, lty = calib.lty)
      }
    }
  }
}


R/print.R
#' Print Method for Monitoring Procedures.
#'
#' Printing objects of class \code{"cointmonitoR"}.
#'
#' @param x [\code{cointmonitoR}]\cr
#'   Object of class \code{"cointmonitoR"}, i.e. the result of
#'   \code{monitorStationarity()} or \code{monitorCointegration()}.
#' @param ... ignored
#' @param digits [\code{numeric}]\cr
#'   Number of significant digits to be used.
#'
#' @return
#'   The invisible \code{x} object.
#'
#' @family cointmonitoR
#'
#' @examples
#' set.seed(42)
#' test = monitorStationarity(rnorm(100), m = 0.5)
#' print(test)
#'
#' x = data.frame(x1 = cumsum(rnorm(200)), x2 = cumsum(rnorm(200)))
#' eps1 = rnorm(200, sd = 2)
#' eps2 = c(eps1[1:100], cumsum(eps1[101:200]))
#' y1 = x$x1 - x$x2 + 10 + eps1
#' y2 = x$x1 - x$x2 + 10 + eps2
#' test1 = monitorCointegration(x = x, y = y1, m = 0.5, model = "FM")
#' print(test1)
#' test2 = monitorCointegration(x = x, y = y2, m = 0.5, model = "FM")
#' print(test2)
#'
#' @export


print.cointmonitoR <- function(x, ..., digits = getOption("digits")) {

  if (!("cointmonitoR" %in% class(x)))
    stop("Argument x must be of type \"cointmonitoR\".")

  firstToCap <- function(x) paste0(toupper(substr(x, 1, 1)), substring(x, 2))

  mon.trend <- firstToCap(x$trend)
  mon.type <- firstToCap(attr(x, "type"))

  cat(paste0("\n", "### Monitoring Procedure for"),
      mon.trend, mon.type, "###\n")

  cat("\nModel:\t")
  cat(paste0("\t",  x$name, "\n"))

  cat("\nParameters:")
  cat(paste0("\t", "m ="), format(x$m[["m.frac"]], digits = digits),
      "(last observation used for calibration:",
      paste0(x$m[["m.index"]], ") \n"))
  cat(ifelse(mon.type != "Stationarity",
             paste0("\t\t", "Model = ",  paste0(x$model, "-OLS"), "  //  ",
                    "Kernel = \"", x$kernel, "\""),
             paste0("\t\t", "Kernel = \"", x$kernel, "\"")),
      " //  Bandwidth =", format(x$bandwidth$number, digits = digits),
      paste0("(", x$bandwidth$name, ")\n"))
  if(!is.null(x$D.options)) {
    cat(paste0("\t\t", "Leads ="), x$D.options$n.lead,
        "/ Lags =", x$D.options$n.lag, "\n")
  }

  cat("\nResults:")
  cat(paste0("\t", "Test Statistic (Hsm) ="),
      format(x$Hsm, digits = digits), "\n")
  cat(paste0("\t\t", "p-Value"),
      ifelse(x$p.value <= 0.01, "<=", ifelse(x$p.value >= 0.1, ">=", "=")),
      format(x$p.value, digits = digits),
      paste0(ifelse(x$p.value > x$sig, "(not ", "("), "significant to"),
      format(x$sig, digits = digits), "level)\n")
  if(x$p.value <= x$sig)
    cat(paste0("\t\t", "Rejection Time ="), x$time, "\n")

  return(invisible(x))
}


R/cointmonitoR.R
#' The cointmonitoR package
#'
#' Consistent Monitoring of Stationarity and Cointegrating Relationships
#'
#' See the vignette:\cr
#' \code{vignette("cointmonitoR")}
#'
#' See the DESCRIPTION:\cr
#' \code{help(package = cointmonitoR)}
#'
#' See the README:\cr
#' \url{https://github.com/aschersleben/cointmonitoR/blob/master/README.md}
#'
#' Open the package documentation page:\cr
#' \code{package?cointmonitoR}
#'
#' Further information and bug reporting:\cr
#' \url{https://github.com/aschersleben/cointmonitoR}
#'
#' @section Functions:
#' \itemize{
#'   \item \code{\link{monitorCointegration}}\cr
#'         This procedure is able to monitor a cointegration model for level or
#'         trend cointegration and returns the corresponding break point, if
#'         available. It is based on parameter estimation on a pre-break
#'         "calibration" period at the beginning of the sample that is known or
#'         assumed to be free of structural change.
#'
#'   \item \code{\link{monitorStationarity}}\cr
#'         This procedure is a special case of \code{monitorCointegration},
#'         since it's able to monitor a one-dimensional vector for level or
#'         trend stationarity.
#'
#'   \item \code{\link[=print.cointmonitoR]{print}}\cr
#'         Print clear results.
#'   \item \code{\link[=plot.cointmonitoR]{plot}}\cr
#'         Plot the test statitics and the values/residuals of a
#'         \code{cointmonitoR} model.
#' }
#'
#' @section Dependencies:
#' This package mainly depends on our
#' \href{https://cran.r-project.org/package=cointReg}{\code{cointReg}} package.
#'
#'
#' @docType package
#' @name cointmonitoR-package
NULL


R/monitorStationarity.R
#' Procedure for Monitoring Level and Trend Stationarity
#'
#' This procedure is able to monitor a one-dimensional vector for level or
#' trend stationarity and returns the corresponding break point, if available.
#' It is based on parameter estimation on a pre-break "calibration" period
#' at the beginning of the sample that is known or assumed to be free of
#' structural change and can be specified exactly via the \code{m} argument
#' (see Details for further information).
#'
#' @param x [\code{numeric} | \code{matrix} | \code{data.frame}]\cr
#'   Data on which to apply the monitoring procedure. If \code{matrix}, it may
#'   have only one row or column, if \code{data.frame} just one column.
#'
#' @param m [\code{numeric(1)}]\cr
#'   Length of calibration period as fraction of the data's length
#'   (between 0.1 and 0.9) or as number of observations (see Details).
#'
#' @param trend [\code{logical}]\cr
#'   Should an intercept and a linear trend be included?
#'   If \code{FALSE} (default), only an intercept is included.
#'
#' @param kernel [\code{character(1)}]\cr
#'   The kernel function to use for calculating the long-run variance.
#'   Default is Bartlett kernel (\code{"ba"}), see Details for alternatives.
#'
#' @param bandwidth [\code{character(1)} | \code{numeric(1)}]\cr
#'   The bandwidth to use for calculating the long-run variance.
#'   Default is Andrews (1991) (\code{"and"}), an alternative is Newey West
#'   (1994) (\code{"nw"}). You can also set the bandwidth manually.
#'
#' @param signif.level [\code{numeric(1)}]\cr
#'   Level of significance (between 0.01 and 0.1).
#'   Detection time will be calculated only if the estimated
#'   p-value is smaller than \code{signif.level}. Default is 0.05.
#'
#' @param return.stats [\code{logical}]\cr
#'   Whether to return all test statistics. Default is \code{TRUE}.
#'
#' @param return.input [\code{logical}]\cr
#'   Whether to return the input data, default is \code{TRUE}.
#'
#' @param check [\code{logical}]\cr
#'   Wheather to check (and if necessary convert) the arguments.
#'   See \code{\link[cointReg]{checkVars}} for further information.
#'
#' @param ...
#'   Arguments passed to \code{\link[cointReg]{getBandwidthNW}} (\code{inter},
#'   \code{weights}), if \code{bandwidth = "nw"}.
#'
#' @details
#'
#' The calibration period can be specified by setting the argument \code{m}
#' to the number of its last observation.
#' The corresponding fraction of the data's length will be calculated
#' automatically. Alternatively you can set \code{m} directly to the fitting
#' fraction value. Attention: The calibration period may become smaller than
#' intended: The last observation is calculated as \code{floor(m * N)}
#' (with \code{N} = length of \code{x}).
#'
#' The kernel that is used for calculating the long-run variance can be
#' one of the following:
#' \itemize{
#'   \item \code{"ba"}: Bartlett kernel
#'   \item \code{"pa"}: Parzen kernel
#'   \item \code{"qs"}: Quadratic Spectral kernel
#'   \item \code{"tr"}: Truncated kernel
#' }
#'
#' @return [\code{cointmonitoR}] object with components:
#' \describe{
#'   \item{\code{Hsm} [\code{numeric(1)}]}{
#'     value of the test statistic}
#'
#'   \item{\code{time} [\code{numeric(1)}]}{
#'     detected time of structural break}
#'
#'   \item{\code{p.value} [\code{numeric(1)}]}{
#'     estimated p-value of the test (between 0.01 and 0.1)}
#'
#'   \item{\code{cv} [\code{numeric(1)}]}{
#'     critical value of the test}
#'
#'   \item{\code{sig} [\code{numeric(1)}]}{
#'     significance level used for the test}
#'
#'   \item{\code{trend} [\code{character(1)}]}{
#'     trend model ("level" or "trend")}
#'
#'   \item{\code{name} [\code{character(1)}]}{
#'     name(s) of data}
#'
#'   \item{\code{m} [\code{list(2)}]}{
#'     list with components:\cr
#'     \code{$m.frac} [\code{numeric(1)}]: calibration period (fraction)\cr
#'     \code{$m.index} [\code{numeric(1)}]: calibration period (length)}
#'
#'   \item{\code{kernel} [\code{character(1)}]}{
#'     kernel function}
#'
#'   \item{\code{bandwidth} [\code{list(2)}]}{
#'     \code{$name} [\code{character(1)}]: bandwidth function (name)\cr
#'     \code{$number} [\code{numeric(1)}]: bandwidth}
#'
#'   \item{\code{statistics} [\code{numeric}]}{
#'     values of test statistics with the same length as data, but \code{NA}
#'     during calibration period (available if \code{return.stats = TRUE})}
#'
#'   \item{\code{input} [\code{numeric} | \code{matrix} | \code{data.frame}]}{
#'     copy of input data (available if \code{return.stats = TRUE})}
#' }
#'
#' @family cointmonitoR
#'
#' @references
#'   \itemize{
#'     \item Wagner, M. and D. Wied (2015): "Monitoring Stationarity and
#'           Cointegration," \emph{Discussion Paper},
#'           \href{http://dx.doi.org/10.2139/ssrn.2624657}{DOI:10.2139/ssrn.2624657}.
#'   }
#'
#' @examples
#' set.seed(1909)
#' x <- rnorm(200)
#' x2 <- c(x[1:100], cumsum(x[101:200]) / 2)
#'
#' # Specify the calibration period
#' # as fraction of the total length of x:
#' monitorStationarity(x, m = 0.25)
#' monitorStationarity(x2, m = 0.465)
#'
#' # Specify the calibration period
#' # by setting its last observation exactly:
#' monitorStationarity(x, m = 50)
#' monitorStationarity(x2, m = 93)
#'
#' @importFrom stats lm predict.lm spline
#'
#' @export


monitorStationarity <- function(x, m = 0.25, trend = FALSE,
                                kernel = c("ba", "pa", "qs", "tr"),
                                bandwidth = c("and", "nw"),
                                signif.level = 0.05, return.stats = TRUE,
                                return.input = TRUE, check = TRUE, ...) {

  x.name <- deparse(substitute(x))

  # check arguments
  if (check) {
    env <- environment()
    cointReg::checkVars(m = m, trend = trend, kernel = kernel,
                        bandwidth = bandwidth, signif.level = signif.level,
                        return.stats = return.stats,
                        return.input = return.input, .env = env)
    x <- cointReg::checkObject(x.stat = x)
  }

  x.T <- nrow(x)

  if (m >= 1 && m <= (0.9 * x.T) && m >= (0.1 * x.T)) {
    m.index <- floor(m)
  } else if (m < 1 && m <= 0.9 && m >= 0.1) {
    m.index <- floor(m * x.T)
  } else {
    stop("m out of possible range.", call. = FALSE)
  }
  m.frac <- m.index / x.T
  if (m != m.frac && m != m.index) {
    warning("Specified parameter m: Observation no. ", m.index,
            " is the end of the calibration period (m = ",
            format(m.frac) ,").", call. = FALSE)
  }

  if (trend) {
    data <- data.frame(x = x, trend = 1:x.T)
    reg <- lm(x ~ 1 + trend, subset = 1:m.index, data = data)
    u <- x - predict.lm(reg, newdata = data)
    ex <- 5
    trend.name <- "trend"
  } else {
    u <- x - mean(x[(1:m.index), ])
    ex <- 3
    trend.name <- "level"
  }

  S <- cumsum(u)
  S.dev <- (S / sqrt(x.T))^2 / x.T
  cumsum.ms <- cumsum(S.dev[(m.index + 1):x.T])
  sum.1m <- sum(S.dev[1:m.index])

  if(!is.numeric(bandwidth)) {
    bw <- cointReg::getBandwidth(u[1:m.index, , drop = FALSE], kernel = kernel,
                                 bandwidth = bandwidth, check = FALSE)
    bandwidth <- switch(bandwidth, and = "Andrews", nw = "Newey-West")
  } else {
    bw <- bandwidth
    bandwidth <- "set by user"
  }
  omega <- cointReg::getLongRunVar(u = u[1:m.index, , drop = FALSE],
                                   kernel = kernel, bandwidth = bw,
                                   demeaning = FALSE, check = FALSE)
  omega <- as.numeric(omega$Omega)

  erg <- (cumsum.ms - sum.1m) / omega
  s <- seq(m.frac, 1, length = x.T - m.index)
  w <- s^ex

  stats.all <- c(rep(NA, m.index), abs(erg / w))
  Hsm <- max(stats.all, na.rm = TRUE)

  cv.table <- cv.stat[[trend.name]]
  cv.which <- (cv.table[, 1] == m.frac)
  if (any(cv.which)) {
    cv.tab <- cv.table[cv.which, 2:5]
  } else {
    cv.tab <- apply(cv.table[, 2:5], 2, function(x) {
      spline(cv.table[, 1], x, xout = m.frac)$y
    })
  }

  p.table <- c(0.1, 0.05, 0.025, 0.01)
  p.value <- spline(cv.tab, p.table, xout = Hsm, method = "natural")$y
  if (p.value < min(p.table)) {
    p.value <- min(p.table)
  } else if (p.value > max(p.table)) {
    p.value <- max(p.table)
  }

  if (signif.level %in% p.table) {
    critical.value <- cv.tab[p.table %in% signif.level]
  } else {
    critical.value <- spline(p.table, cv.tab, xout = signif.level,
                             method = "natural")$y
  }
  critical.value <- as.numeric(critical.value)

  if (Hsm > critical.value) {
    rej.time <- min(which(stats.all > critical.value))
  } else {
    rej.time <- Inf
  }

  out <- list(Hsm = Hsm, time = rej.time,
              p.value = p.value, cv = critical.value,
              sig = signif.level, trend = trend.name,
              name = x.name, m = list(m.frac = m.frac, m.index = m.index),
              kernel = kernel, bandwidth = list(name = bandwidth, number = bw))

  if(return.stats) {
    out <- c(out, statistics = list(stats.all))
  }

  if(return.input) {
    out <- c(out, input = list(x))
  }

  class(out) <- "cointmonitoR"
  attr(out, "type") <- "stationarity"
  return(out)
}
